posterior_averaging/main/federated_trainer.py

```diff
@@ -0,0 +1,295 @@
+  flags.DEFINE_integer(
+      'so_lr_num_validation_examples', 10000, 'Number of examples '
+      'to use from test set for per-round validation.')
+  flags.DEFINE_integer('so_lr_max_elements_per_user', 1000,
+                       'Max number of training '
+                       'sentences to use per user.')
+
+FLAGS = flags.FLAGS
+
+TASK_FLAGS = collections.OrderedDict(
+    cifar100=cifar100_flags,
+    emnist_cr=emnist_cr_flags,
+    shakespeare=shakespeare_flags,
+    stackoverflow_nwp=so_nwp_flags,
+    stackoverflow_lr=so_lr_flags)
+
+
+def _write_hparam_flags():
+  """Creates an ordered dictionary of hyperparameter flags and writes to CSV."""
+  hparam_dict = utils_impl.lookup_flag_values(shared_flags)
+
+  # Update with optimizer flags corresponding to the chosen optimizers.
+  opt_flag_dict = utils_impl.lookup_flag_values(optimizer_flags)
+  opt_flag_dict = optimizer_utils.remove_unused_flags('client', opt_flag_dict)
+  opt_flag_dict = optimizer_utils.remove_unused_flags('server', opt_flag_dict)
+  hparam_dict.update(opt_flag_dict)
+
+  # Update with task-specific flags.
+  task_name = FLAGS.task
+  if task_name in TASK_FLAGS:
+    task_hparam_dict = utils_impl.lookup_flag_values(TASK_FLAGS[task_name])
+    hparam_dict.update(task_hparam_dict)
+
+  results_dir = os.path.join(FLAGS.root_output_dir, 'results',
+                             FLAGS.experiment_name)
+  utils_impl.create_directory_if_not_exists(results_dir)
+  hparam_file = os.path.join(results_dir, 'hparams.csv')
+  utils_impl.atomic_write_series_to_csv(hparam_dict, hparam_file)
+
+
+def main(argv):
+  if len(argv) > 1:
+    raise app.UsageError('Expected no command-line arguments, '
+                         'got: {}'.format(argv))
+
+  client_optimizer_fn = optimizer_utils.create_optimizer_fn_from_flags('client')
+  server_optimizer_fn = optimizer_utils.create_optimizer_fn_from_flags('server')
+
+  client_lr_schedule = optimizer_utils.create_lr_schedule_from_flags('client')
+  server_lr_schedule = optimizer_utils.create_lr_schedule_from_flags('server')
+
+  client_mixedin_schedule_fn = fed_pa_schedule.create_mixin_check_fn(
+      name=FLAGS.client_mixin_check_scheme,
+      num_mixin_epochs=FLAGS.client_mixin_epochs_per_round,
+      start_round=FLAGS.client_mixin_check_start_round)
+  client_update_delta_fn = fed_pa_schedule.create_update_delta_fn(
+      name=FLAGS.client_update_delta_scheme, rho=FLAGS.client_shrinkage_rho)
+
+  def iterative_process_builder(
+      model_fn: Callable[[],
+                         tff.learning.Model]) -> tff.templates.IterativeProcess:
+    """Creates an iterative process using a given TFF `model_fn`.
+
+    Args:
+      model_fn: A no-arg function returning a `tff.learning.Model`.
+
+    Returns:
+      A `tff.templates.IterativeProcess`.
+    """
+    if FLAGS.task == 'shakespeare' or FLAGS.task == 'stackoverflow_nwp':
+
+      def client_weight_fn(local_outputs):
```

Comment:
I've seen little to no difference between 'num_tokens' and the default example weighting, so if you'd like to reduce complexity you could remove these two lines.